webknossos/webknossos/geometry/nd_bounding_box.py

```diff
@@ -0,0 +1,507 @@
+        index: List[int] = [1, 2, 3]
+
+        if "axisOrder" in bbox:
+            axes = list(bbox["axisOrder"].keys())
+            index = [bbox["axisOrder"][axis] for axis in axes]
+
+            if "additionalAxes" in bbox:
+                assert (
+                    "axisOrder" in bbox
+                ), "If there are additionalAxes an axisOrder needs to be given."
+                for axis in bbox["additionalAxes"]:
+                    topleft.append(axis["bounds"][0])
+                    size.append(axis["bounds"][1] - axis["bounds"][0])
+                    axes.append(axis["name"])
+                    index.append(axis["index"])
+
+        return cls(topleft, size, axes, index)
+
+    def to_wkw_dict(self) -> dict:
+        topleft = [None, None, None]
+        width, height, depth = None, None, None
+        additional_axes = []
+        for i, axis in enumerate(self.axes):
+            if axis == "x":
+                topleft[0] = self.topleft[i]
+                width = self.size[i]
+            elif axis == "y":
+                topleft[1] = self.topleft[i]
+                height = self.size[i]
+            elif axis == "z":
+                topleft[2] = self.topleft[i]
+                depth = self.size[i]
+            else:
+                additional_axes.append(
+                    {
+                        "name": axis,
+                        "bounds": [self.topleft[i], self.bottomright[i]],
+                        "index": self.index[i],
+                    }
+                )
+
+        return {
+            "topLeft": topleft,
+            "width": width,
+            "height": height,
+            "depth": depth,
+            "additionalAxes": additional_axes,
+        }
+
+    def to_config_dict(self) -> dict:
+        return {
+            "topleft": self.topleft.to_list(),
+            "size": self.size.to_list(),
+            "axes": self.axes,
+        }
+
+    def to_checkpoint_name(self) -> str:
+        return f"{'_'.join(str(element) for element in self.topleft)}_{'_'.join(str(element) for element in self.size)}"
+
+    def __repr__(self) -> str:
+        return f"NDBoundingBox(topleft={self.topleft.to_tuple()}, size={self.size.to_tuple()}, axes={self.axes})"
+
+    def __str__(self) -> str:
+        return self.__repr__()
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, NDBoundingBox):
+            self._check_compatibility(other)
+            return self.topleft == other.topleft and self.size == other.size
+
+        raise NotImplementedError()
+
+    def __len__(self) -> int:
+        return len(self.axes)
+
+    def get_shape(self, axis_name: str) -> int:
+        try:
+            index = self.axes.index(axis_name)
+            return self.size[index]
+        except ValueError as err:
+            raise ValueError(
+                f"Axis {axis_name} doesn't exist in NDBoundingBox."
+            ) from err
+
+    def get_3d(self, attr_name: str) -> Vec3Int:
```

Comment:
I think, this method is too generic and (seemingly) powerful to be exposed on this class. `attr_name` only works for very specific strings and we should not put this complexity on external users.
You probably wanted to avoid code duplication, but I'd only keep this function as a helper and suggest something like this:

```python
def _get_attr_xyz(self, attr_name: str) -> Vec3Int:
	# private. old get_3d code
	...

def _set_attr_xyz(self, attr_name: str, value: Vec3IntLike) -> VecInt:
	# private old set_3d code
	...

@property
def topleft_xyz(self, ...):
	return self._get_attr_xyz("topleft")
	...

@property
def size_xyz(self, ...):
	...

@topleft_xyz.setter
def topleft_xyz(self, ...):
	...
```

Reply:
I like the idea of having a getter for each attribute. The setter is not possible because the NDBoundingBox class is immutable. It wasn't a setter in the first place, but a getter with even worse naming. I think it should look somewhat like that:
```python
def _get_attr_xyz(self, attr_name: str) -> Vec3Int:
	# private. old get_3d code
	...

def _get_attr_with_replaced_xyz(self, attr_name: str, new_xyz: Vec3IntLike) -> VecInt:
	# private old set_3d code
	...

@property
def topleft_xyz(self) -> Vec3Int:
	return self._get_attr_xyz("topleft")

@property
def size_xyz(self) -> Vec3Int:
	...

def topleft_with_xyz(self, new_xyz: Vec3IntLike) -> VecInt:
	...
	
def size_with_xyz(self, new_xyz: Vec3IntLike) -> VecInt:
	...
```

Reply:
with_* functions sound good. I think the `with_` should be a prefix, though.

> ```python
> def with_topleft_xyz(self, new_xyz: Vec3IntLike) -> VecInt:
> 	...
> 	
> def with_size_xyz(self, new_xyz: Vec3IntLike) -> VecInt:
> 	...
> ```

Reply:
> I think the with_ should be a prefix, though.

I think, this would only make sense when the method would return instances of the same class (i.e., new BoundingBoxNd instances). However, it returns the requested attribute with changed values. This is why I think, `topleft_with_xyz` is better.

Reply:
Done.